### ISSUE TITLE
[Build 2023122703-4.2] sys-usb panics on Framework 13 AMD, with 6.1.62 kernel

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#8794]
## Observation

openQA test in scenario qubesos-4.2-kernel-x86_64-install_default_upload@hw12 fails in
[firstboot](https://openqa.qubes-os.org/tests/88946/modules/firstboot/steps/84)

```
[2023-12-27 12:36:02] [    2.430791] thunderbolt 0000:00:0b.0: Failed to create MSI-X! ret=-19!
[2023-12-27 12:36:02] [    2.430803] list_del corruption, ffff99605255c0c0->next is LIST_POISON1 (dead000000000100)
[2023-12-27 12:36:02] [    2.430815] ------------[ cut here ]------------
[2023-12-27 12:36:02] [    2.430821] kernel BUG at lib/list_debug.c:53!
[2023-12-27 12:36:02] [    2.430829] invalid opcode: 0000 [#1] PREEMPT SMP NOPTI
[2023-12-27 12:36:02] [    2.430835] CPU: 1 PID: 296 Comm: (udev-worker) Not tainted 6.1.62-1.qubes.fc37.x86_64 #1
[2023-12-27 12:36:02] [    2.430843] Hardware name: Xen HVM domU, BIOS 4.17.2 12/10/2023
[2023-12-27 12:36:02] [    2.430849] RIP: 0010:__list_del_entry_valid.cold+0x5c/0x6f
[2023-12-27 12:36:02] [    2.430859] Code: e8 1d 7a fd ff 0f 0b 48 89 fe 48 89 ca 48 c7 c7 98 60 9e bb e8 09 7a fd ff 0f 0b 48 89 fe 48 c7 c7 60 60 9e bb e8 f8 79 fd ff <0f> 0b 48 89 fe 48 c7 c7 30 60 9e bb e8 e7 79 fd ff 0f 0b 48 8b 04
[2023-12-27 12:36:02] [    2.430875] RSP: 0018:ffffb8bc402e3a40 EFLAGS: 00010246
[2023-12-27 12:36:02] [    2.430882] RAX: 000000000000004e RBX: ffff99605255c0c0 RCX: 0000000000000000
[2023-12-27 12:36:02] [    2.430889] RDX: 0000000000000000 RSI: ffffffffbb9cb271 RDI: 00000000ffffffff
[2023-12-27 12:36:02] [    2.430896] RBP: ffff99605255c0c0 R08: 0000000000000000 R09: ffffb8bc402e38e0
[2023-12-27 12:36:02] [    2.430903] R10: 0000000000000003 R11: ffffffffbbd471c8 R12: ffff99605255c0d0
[2023-12-27 12:36:02] [    2.430910] R13: ffffb8bc40275000 R14: ffff996041c39000 R15: 0000000000000010
[2023-12-27 12:36:02] [    2.430918] FS:  000072e524a27940(0000) GS:ffff996057100000(0000) knlGS:0000000000000000
[2023-12-27 12:36:02] [    2.430926] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[2023-12-27 12:36:02] [    2.430932] CR2: 00005f8d83b9d000 CR3: 000000000d154000 CR4: 0000000000750ee0
[2023-12-27 12:36:02] [    2.430940] PKRU: 55555554
[2023-12-27 12:36:02] [    2.430943] Call Trace:
[2023-12-27 12:36:02] [    2.430947]  <TASK>
[2023-12-27 12:36:02] [    2.430951]  ? show_trace_log_lvl+0x1d3/0x2ef
[2023-12-27 12:36:02] [    2.430958]  ? show_trace_log_lvl+0x1d3/0x2ef
[2023-12-27 12:36:02] [    2.430964]  ? show_trace_log_lvl+0x1d3/0x2ef
[2023-12-27 12:36:02] [    2.430970]  ? xen_free_irq+0x92/0x100
[2023-12-27 12:36:02] [    2.430976]  ? __die_body.cold+0x8/0xd
[2023-12-27 12:36:02] [    2.430981]  ? die+0x2a/0x50
[2023-12-27 12:36:02] [    2.430987]  ? do_trap+0xc5/0x110
[2023-12-27 12:36:02] [    2.430992]  ? __list_del_entry_valid.cold+0x5c/0x6f
[2023-12-27 12:36:02] [    2.430998]  ? do_error_trap+0x6a/0x90
[2023-12-27 12:36:02] [    2.431002]  ? __list_del_entry_valid.cold+0x5c/0x6f
[2023-12-27 12:36:02] [    2.431008]  ? exc_invalid_op+0x4c/0x60
[2023-12-27 12:36:02] [    2.431015]  ? __list_del_entry_valid.cold+0x5c/0x6f
[2023-12-27 12:36:02] [    2.431020]  ? asm_exc_invalid_op+0x16/0x20
[2023-12-27 12:36:02] [    2.431028]  ? __list_del_entry_valid.cold+0x5c/0x6f
[2023-12-27 12:36:02] [    2.431034]  xen_free_irq+0x92/0x100
[2023-12-27 12:36:02] [    2.431039]  xen_destroy_irq+0x6c/0x120
[2023-12-27 12:36:02] [    2.431044]  xen_teardown_msi_irqs+0x3b/0x70
[2023-12-27 12:36:02] [    2.431051]  msi_domain_free_irqs_descs_locked+0x1b/0x40
[2023-12-27 12:36:02] [    2.431058]  pci_msi_teardown_msi_irqs+0x3e/0x40
[2023-12-27 12:36:02] [    2.431065]  __pci_enable_msix_range.part.0+0x2d4/0x490
[2023-12-27 12:36:02] [    2.431072]  pci_alloc_irq_vectors_affinity+0xa9/0x110
[2023-12-27 12:36:02] [    2.431079]  nhi_probe+0x1bf/0x510 [thunderbolt]
[2023-12-27 12:36:02] [    2.431096]  local_pci_probe+0x41/0x80
[2023-12-27 12:36:02] [    2.431101]  pci_call_probe+0x54/0x160
[2023-12-27 12:36:02] [    2.431106]  pci_device_probe+0x7c/0x100
[2023-12-27 12:36:02] [    2.431113]  ? driver_sysfs_add+0x71/0xd0
[2023-12-27 12:36:02] [    2.431118]  really_probe+0xde/0x380
[2023-12-27 12:36:02] [    2.431123]  ? pm_runtime_barrier+0x50/0x90
[2023-12-27 12:36:02] [    2.431128]  __driver_probe_device+0x78/0x120
[2023-12-27 12:36:02] [    2.431134]  driver_probe_device+0x1f/0x90
[2023-12-27 12:36:02] [    2.431138]  __driver_attach+0xce/0x1c0
[2023-12-27 12:36:02] [    2.431143]  ? __device_attach_driver+0x110/0x110
[2023-12-27 12:36:02] [    2.431148]  bus_for_each_dev+0x87/0xd0
[2023-12-27 12:36:02] [    2.431154]  bus_add_driver+0x1ae/0x200
[2023-12-27 12:36:02] [    2.431159]  driver_register+0x89/0xe0
[2023-12-27 12:36:02] [    2.431163]  nhi_init+0x5c/0x1000 [thunderbolt]
[2023-12-27 12:36:02] [    2.431177]  ? 0xffffffffc074b000
[2023-12-27 12:36:02] [    2.431181]  do_one_initcall+0x59/0x230
[2023-12-27 12:36:02] [    2.431188]  do_init_module+0x4a/0x1f0
[2023-12-27 12:36:02] [    2.431194]  __do_sys_finit_module+0xac/0x120
[2023-12-27 12:36:02] [    2.431200]  do_syscall_64+0x5b/0x80
[2023-12-27 12:36:02] [    2.431206]  ? do_syscall_64+0x67/0x80
[2023-12-27 12:36:02] [    2.431210]  entry_SYSCALL_64_after_hwframe+0x64/0xce
[2023-12-27 12:36:02] [    2.431217] RIP: 0033:0x72e524d2cb4d
[2023-12-27 12:36:02] [    2.431222] Code: c3 66 2e 0f 1f 84 00 00 00 00 00 66 90 f3 0f 1e fa 48 89 f8 48 89 f7 48 89 d6 48 89 ca 4d 89 c2 4d 89 c8 4c 8b 4c 24 08 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d 8b 92 0c 00 f7 d8 64 89 01 48
[2023-12-27 12:36:02] [    2.431237] RSP: 002b:00007ffc467850c8 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
[2023-12-27 12:36:02] [    2.431245] RAX: ffffffffffffffda RBX: 00005f8d8390ce00 RCX: 000072e524d2cb4d
[2023-12-27 12:36:02] [    2.431252] RDX: 0000000000000000 RSI: 000072e52524307d RDI: 0000000000000006
[2023-12-27 12:36:02] [    2.431260] RBP: 00007ffc46785180 R08: 0000000000000000 R09: 00007ffc46785130
[2023-12-27 12:36:02] [    2.431266] R10: 0000000000000006 R11: 0000000000000246 R12: 000072e52524307d
[2023-12-27 12:36:02] [    2.431273] R13: 0000000000020000 R14: 00005f8d833ddb10 R15: 00005f8d833dda00
[2023-12-27 12:36:02] [    2.431281]  </TASK>
[2023-12-27 12:36:02] [    2.431284] Modules linked in: intel_rapl_msr intel_rapl_common crct10dif_pclmul crc32_pclmul crc32c_intel polyval_clmulni polyval_generic ghash_clmulni_intel sha512_ssse3 pcspkr thunderbolt(+) drm_vram_helper drm_ttm_helper ttm xhci_pci xhci_pci_renesas ehci_pci ata_generic xhci_hcd serio_raw i2c_piix4 ehci_hcd pata_acpi xen_scsiback target_core_mod xen_netback xen_privcmd xen_gntdev xen_gntalloc xen_blkback xen_evtchn fuse loop overlay xen_blkfront
[2023-12-27 12:36:02] [    2.431335] ---[ end trace 0000000000000000 ]---
[2023-12-27 12:36:02] [    2.431340] RIP: 0010:__list_del_entry_valid.cold+0x5c/0x6f
[2023-12-27 12:36:02] [    2.431347] Code: e8 1d 7a fd ff 0f 0b 48 89 fe 48 89 ca 48 c7 c7 98 60 9e bb e8 09 7a fd ff 0f 0b 48 89 fe 48 c7 c7 60 60 9e bb e8 f8 79 fd ff <0f> 0b 48 89 fe 48 c7 c7 30 60 9e bb e8 e7 79 fd ff 0f 0b 48 8b 04
[2023-12-27 12:36:02] [    2.431362] RSP: 0018:ffffb8bc402e3a40 EFLAGS: 00010246
[2023-12-27 12:36:02] [    2.431368] RAX: 000000000000004e RBX: ffff99605255c0c0 RCX: 0000000000000000
[2023-12-27 12:36:02] [    2.431375] RDX: 0000000000000000 RSI: ffffffffbb9cb271 RDI: 00000000ffffffff
[2023-12-27 12:36:02] [    2.431381] RBP: ffff99605255c0c0 R08: 0000000000000000 R09: ffffb8bc402e38e0
[2023-12-27 12:36:02] [    2.431388] R10: 0000000000000003 R11: ffffffffbbd471c8 R12: ffff99605255c0d0
[2023-12-27 12:36:02] [    2.431396] R13: ffffb8bc40275000 R14: ffff99[2023-12-27 12:36:02] 6041c39000 R15: 0000000000000010
[2023-12-27 12:36:02] [    2.431404] FS:  000072e524a27940(0000) GS:ffff996057100000(0000) knlGS:0000000000000000
[2023-12-27 12:36:02] [    2.431411] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[2023-12-27 12:36:02] [    2.431417] CR2: 00005f8d83b9d000 CR3: 000000000d154000 CR4: 0000000000750ee0
[2023-12-27 12:36:02] [    2.431424] PKRU: 55555554
[2023-12-27 12:36:02] [    2.431426] Kernel panic - not syncing: Fatal exception
[2023-12-27 12:36:02] [    2.431455] Kernel Offset: 0x39000000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
```

## Test suite description
Default installation, preserve disk for further tests

## Reproducible

Fails since (at least) Build [2023120510-4.2](https://openqa.qubes-os.org/tests/88806)

## Expected result

Last good: (unknown) (or more recent)

## Further details

Always latest result in this scenario: [latest](https://openqa.qubes-os.org/tests/latest?arch=x86_64&distri=qubesos&flavor=kernel&machine=hw12&test=install_default_upload&version=4.2)

**Comment by marmarek** [comment]
And with kernel-latest, the whole host (dom0?) crashes when starting sys-net/sys-usb: https://openqa.qubes-os.org/tests/89014
No console log, sadly.

**Comment by euidzero** [comment]
It  hard reboots here if something is connected to the USB4/Thunderbolt port when sys-usb starts. Starting without any connected device works.

**Comment by euidzero** [comment]
Still there with kernel 6.9.4-1.qubes.fc37.x86_64

**Comment by euidzero** [comment]
And rebooting sys-usb (with or without anything connected) does the same. Maybe somewhat related to passthrough/FLR
 For every USB controller I see in dmesg : 
```
pciback 0000:c3:00.3: xen-pciback: Driver tried to write to a read-only configuration space field at offset 0x2a6, size 2. This may be harmless, but if you have problems with your device:
               1) see permissive attribute in sysfs
               2) report problems to the xen-devel mailing list along with details of your device obtained from lspci.
```

I 'm experiencing with ` no-strict-reset=True|False` on USB4 controllers but this does not seems to help a lot.

**Comment by Ofenhed** [comment]
This happens because of how `qubes-pciback` unbinds devices. I cannot claim to fully understand the issue, but it seems that it breaks if you try to bind a USB controller which has a (USB3?) device connected, while not having unbound other USB controllers on the same PCI bus. I also had some issues with using `pciback/new_slot` and `pciback/bind`

This can be made to work by unbinding all devices before rebinding any of them, and by using `drive_override` and `drivers_probe`. [Here's a working snippet doing just that](https://github.com/Ofenhed/qubes-salts/blob/a3bb34a3eb7d3630daf96c8d0773a27039cbffaf/hide-all-usb-after-decrypt.sls#L156-L181).

**Comment by marmarek** [comment]
> This happens because of how `qubes-pciback` unbinds devices.

If `qubes-pciback` is expected to hide USB controllers (`rd.qubes.hide_all_usb` is on the kernel cmdline), it does it before normal USB drivers are even loaded, so there is no unbinding involved (devices are not bound to the driver yet). But if that option is not used (for example when you need USB keyboard to enter disk passphrase), then unbinding is done at sys-usb startup by libvirt. And it does it very similar to your snippet, using `driver_override`: https://gitlab.com/libvirt/libvirt/-/blob/master/src/util/virpci.c?ref_type=heads#L1580-1610

**Comment by Ofenhed** [comment]
@marmarek but that code has the same problem. It rebinds before the full PCI bus is unbound. 

Take this with a grain of salt, it was quite a long time since I debugged this, but switching to driver_override and drivers_probe was not enough.

This caused the crash:
```
Unbind 03:00.3 // first USB-C port
Probe 03:00.3
Unbind 03:00.4 // second USB-C port
Probe 03:00.4
...
Boot sys-usb
```

This did not:
```
Unbind 03:00.3
Unbind 03:00.4
...
Probe 03:00.3
Probe 03:00.4
...
Boot sys-usb
```

I moved my rebinds to systemd for convenience, but I changed the rebinding to "atomically" unbind *all* the controllers before i probed the first controller. I don't recall whether I tried it in cmdline, but I would assume that the snippet (with some modifications for error handling) would work as well in cmdline and solve this crash.

**Comment by Ofenhed** [comment]
You were right, I had gotten it mixed up. The issue does not arise when `qubes-pciback` is running with `rd.qubes.hide_all_usb`, but it arises when that option is omitted, because libvirt unbinds and rebinds non-atomically.
Is this a separate issue then, and the bug I described only part of #9291? Since it's actually either a hardware quirk or a bug in libvirt, I guess I should contact them...

**Comment by marmarek** [comment]
In either case it is a Linux kernel issue, it should not crash regardless of the unbind order.

**Comment by Ofenhed** [comment]
Well... I'm in way over my head here... I somewhat know in an intuitive level what happens here, and how to workaround, but I haven't the first clue on how to turn this into a useful bug report for Linux...

**Comment by marmarek** [comment]
Some further info on this (or very similar) issue with xen unstable: https://lore.kernel.org/xen-devel/aIDXIqA4L7wcJH2T@mail-itl/

**Comment by superm1** [comment]
> Some further info on this (or very similar) issue with xen unstable: https://lore.kernel.org/xen-devel/aIDXIqA4L7wcJH2T@mail-itl/

Just a guess; could you be tripping a page fault from the IOMMU configuration by accessing the controller this way?

**Comment by marmarek** [comment]
Maybe, but there would be more to that. IOMMU faults are gracefully handled by Xen, with appropriate log messages (which I haven't seen here). And it isn't fatal for the whole system (if anything, may result in a device in a weird state, but not hard reset). Note also that at this stage of VM startup, no DMA should be issued yet - it's quite early in the SeaBIOS when it is just discovering device capabilities, not enabling controller yet.

**Comment by superm1** [comment]
But the point that I'm worried about is the iommu was set up preboot. If you didn't reset at this point you might not have a handler ready if it's still enabled but not configured in Xen.

**Comment by marmarek** [comment]
Xen and dom0 start just fine, system crashes when starting sys-usb domU

**Comment by superm1** [comment]
Considering the problem you're seeing (sync flood); are you on the latest FW from Framework already?

**Comment by marmarek** [comment]
Yes, 3.09.

**Comment by superm1** [comment]
When passed through to the VM how are you handling ACPI power state management?  Are you passing through the matching ACPI objects to the VM somehow?  IE I'm wondering if you have an initial power state mismatch that they're in D3cold when starting the VM but without the correct ACPI objects you don't properly transition to D0.

**Comment by marmarek** [comment]
That part is completely ignored, no ACPI related to PCI devices are shared into VM. But still, it shouldn't crash the whole host, no? At worst (in theory), should result in a non-functional controller... Do you know which ACPI objects are relevant there (on Framework specifically)? PCI passthrough of USB controllers work just fine without messing with ACPI on other systems (including other AMD)...

**Comment by superm1** [comment]
I mean accessing registers of a device that's in D3cold is never going to be safe.  There are a lot of important methods on the ACPI object for the XHCI controller (IE `_PS0`, `_PS3`, `_REG`).  If you look at some of them they work of stateful region of system memory `VARR`.  
```
            Field (VARM, AnyAcc, NoLock, Preserve)
            {
                VARR,   32
            }

            Local1 = VARR /* \_SB_.PCI0.GPPA.XHC1.RPRM.VARR */
```
I don't know for sure; but I would not put it out of the question that the low level platform firmware has an interrupt triggered on writes to this region that controls whether the IP is powered up.  My hypothesis is that trying to change the power state without doing the matching _PS0 AML means that the IP stays powered down and you are seeing what happens when you try to access IP that is powered down.  

To confirm this is the cause or not can you ensure that the hypervisor (which has access to the ACPI resources) transitions the device to D0 and it **stays** there when detaching (presumably using something like virtio for passthrough)?

**Comment by tasket** [comment]
I'm not sure if this issue is dedicated to Framework, but this panic/reboot issue has existed for a long time on my Lenovo T14 Ryzen system and is still there with Qubes R4.3 rc3.  Workaround is to remove the auto-start for sys-usb and unplug any devices before starting that VM.

**Comment by alimirjamali** [comment]
Some workaround on this issue posted by \@ahf on [Mastodon](https://mastodon.social/@ahf/115763116884499624).

> I've been using Qubes 4.3 in the RC phase on my Framework 13 laptop. I usually go with kernel-latest, but started seeing a bootup freeze right after the amdgpu driver was initialized. I solved this by manually blocking the `amdxdna` driver from loading. You can do this either with the `module_blacklist` kernel argument or via `/etc/modprobe.d/` and re-generating your grub environment with `grub2-mkconfig`.
